### ISSUE TITLE
Do evaluation and save checkpoint while using multi TPUs/GPUs

### ISSUE THREAD

**Issue by wyhhyw123** [pytorch/xla#3468]
## ❓ Questions and Help
Hello, I am curious about the evalution method in `xla/test/test_train_mp_mnist.py`. There is a operation 'mesh_reduce'  in 'test_loop_fn', i don't understand why the operation is needed. Does this mean both the evaluation result and model parameters derived from different TPUs/GPUs were not the same? If the answer is yes, so how to save the model ckpt? 

In my mind, model parameters in each GPU will reach the same after the operation 'all_reduce' while using torch, so we usually save the model ckpt in GPU0. 

![image](https://user-images.githubusercontent.com/63039413/161373468-ea0c8f39-b82b-4a8d-aec6-5dac8588f103.png)

**Comment by taylanbil** [comment]
When this workload runs on a v3-8 for example, it runs in a "data parallel" way on 8 different devices.

What this means is that there are 8 processes, each containing the **exact same** copy of the model. However, the data that each process sees is different (if that was also the same, there would be no reason to run the same job on all devices, that would be just duplicate work.)

Since the input data is different (check [this part of code](https://github.com/pytorch/xla/blob/master/test/test_train_mp_imagenet.py#L155) and how `num_replicas` and `rank` are used), the accuracy values in line 48 in the screenshot are different. In order to report one single accuracy at the end, one would need to average out the 8 accuracy values across devices. The averaged accuracy is the global accuracy for the whole training job.

So, to summarize, the last sentence you wrote *is* correct, the all_reduce guarantees that each device holds the same exact model weights in a data parallel workload. Given that, the need for mesh_reduce doesn't come from model weights being different, but that the input data is different across different devices / processes, and to get the global accuracy as the single number, one needs to reduce.

Closing this for now, please reopen if needed.

**Comment by jotron** [comment]
Hi, what you said makes sense to me for *xla/test/test_train_mp_imagenet.py*. But in *xla/test/test_train_mp_mnist.py*  the data seen is the same as I understand it, see [the code](https://github.com/pytorch/xla/blob/master/test/test_train_mp_mnist.py#L104).
But still when I print the accuracy for each core I still get very slightly different values.

```
...
Epoch 1 train end 14:23:36
---5 correct=9253, total_samples=10000
---7 correct=9251, total_samples=10000
---4 correct=9242, total_samples=10000
---0 correct=9245, total_samples=10000
---2 correct=9247, total_samples=10000
---6 correct=9242, total_samples=10000
---3 correct=9256, total_samples=10000
---1 correct=9258, total_samples=10000
Epoch 1 test end 14:23:41, Accuracy=92.49
...
```

So does this mean that the models across cores are slightly different?

**Comment by JackCaoG** [comment]
I reopen the previous issue, we can follow up with the syncbn op lowering.
